wasm/derive/src/validator/token.rs

```diff
@@ -0,0 +1,179 @@
+
+use super::*;
+
+/// [`derive_token`](crate::derive_token()) macro implementation
+pub fn impl_derive_token(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let ident = input.ident;
+
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+    let syn::Data::Struct(syn::DataStruct { fields, .. }) = input.data else {
+        panic!("`Token` can be derived only for structs");
+    };
+    let extracted_fields = match fields {
+        syn::Fields::Named(syn::FieldsNamed { named, .. }) => named,
+        syn::Fields::Unit => syn::punctuated::Punctuated::default(),
+        _ => panic!("`Token` can be derived only for structs with named fields or unit structs"),
+    };
+
+    let impl_token = impl_token(
+        &impl_generics,
+        &ty_generics,
+        where_clause,
+        &ident,
+        &extracted_fields,
+    );
+    let impl_try_from_permission_token = impl_try_from_permission_token(
+        &impl_generics,
+        &ty_generics,
+        where_clause,
+        &ident,
+        &extracted_fields,
+    );
+
+    quote! {
+        #impl_token
+        #impl_try_from_permission_token
+    }
+    .into()
+}
+
+fn impl_token(
+    impl_generics: &syn::ImplGenerics<'_>,
+    ty_generics: &syn::TypeGenerics<'_>,
+    where_clause: Option<&syn::WhereClause>,
+    ident: &syn::Ident,
+    fields: &syn::punctuated::Punctuated<syn::Field, syn::token::Comma>,
+) -> proc_macro2::TokenStream {
+    use heck::ToSnakeCase as _;
+
+    let definition_id = proc_macro2::Literal::string(&ident.to_string().to_snake_case());
+    let permission_token_conversion_code = permission_token_conversion(fields);
+
+    quote! {
+        impl #impl_generics ::iroha_wasm::validator::traits::Token for #ident #ty_generics
+        #where_clause
+        {
+            fn definition_id() -> ::iroha_wasm::data_model::permission::token::Id {
+                ::iroha_wasm::parse!(
+                    #definition_id as <
+                        ::iroha_wasm::data_model::permission::token::Definition
+                        as
+                        ::iroha_wasm::data_model::prelude::Identifiable
+                    >::Id
+                )
+            }
+
+            fn is_owned_by(
+                &self,
+                account_id: &<
+                    ::iroha_wasm::data_model::prelude::Account
+                    as
+                    ::iroha_wasm::data_model::prelude::Identifiable
+                >::Id
+            ) -> bool {
+                use ::iroha_wasm::Execute as _;
+
+                let permission_token = #permission_token_conversion_code;
+
+                ::iroha_wasm::data_model::prelude::QueryBox::from(
+                    ::iroha_wasm::data_model::prelude::DoesAccountHavePermissionToken {
+                        account_id: account_id.clone().into(),
+                        permission_token,
+                    }
```

Comment:
please use constructor this here as well:
```suggestion
                    ::iroha_wasm::data_model::prelude::DoesAccountHavePermissionToken::new(
                        account_id.clone().into(),
                        permission_token,
                    }
```